src/Interpreters/InterpreterCheckQuery.cpp

```diff
@@ -3,77 +3,211 @@
+
+    void consume(Chunk chunk) override
+    {
+        if (result_value == 0)
+            return;
+
+        auto columns = chunk.getColumns();
+        if (columns.size() != 3)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Wrong number of columns: {}", columns.size());
+
+        const auto * col = checkAndGetColumn<ColumnUInt8>(columns[1].get());
+        for (size_t i = 0; i < col->size(); ++i)
+        {
+            if (col->getElement(i) == 0)
+            {
+                result_value = 0;
+                return;
+            }
+        }
+    }
+
+    Chunk generate() override
+    {
+        if (is_value_emitted.exchange(true))
+            return {};
+        auto block = getSingleValueBlock(result_value);
+        return Chunk(block.getColumns(), block.rows());
+    }
+
+private:
+    std::atomic<UInt8> result_value{1};
+    std::atomic_bool is_value_emitted{false};
+};
+
 }
 
+InterpreterCheckQuery::InterpreterCheckQuery(const ASTPtr & query_ptr_, ContextPtr context_)
+    : WithContext(context_)
+    , query_ptr(query_ptr_)
+{
+}
 
 BlockIO InterpreterCheckQuery::execute()
 {
     const auto & check = query_ptr->as<ASTCheckQuery &>();
-    auto table_id = getContext()->resolveStorageID(check, Context::ResolveOrdinary);
+    const auto & context = getContext();
+    auto table_id = context->resolveStorageID(check, Context::ResolveOrdinary);
 
-    getContext()->checkAccess(AccessType::SHOW_TABLES, table_id);
-    StoragePtr table = DatabaseCatalog::instance().getTable(table_id, getContext());
-    auto check_results = table->checkData(query_ptr, getContext());
+    context->checkAccess(AccessType::SHOW_TABLES, table_id);
+    StoragePtr table = DatabaseCatalog::instance().getTable(table_id, context);
 
-    Block block;
-    if (getContext()->getSettingsRef().check_query_single_value_result)
-    {
-        bool result = std::all_of(check_results.begin(), check_results.end(), [] (const CheckResult & res) { return res.success; });
-        auto column = ColumnUInt8::create();
-        column->insertValue(static_cast<UInt64>(result));
-        block = Block{{std::move(column), std::make_shared<DataTypeUInt8>(), "result"}};
-    }
-    else
+    auto check_data_tasks = table->getCheckTaskList(query_ptr, context);
+
+    const auto & settings = context->getSettingsRef();
+
+    BlockIO res;
     {
-        auto block_structure = getBlockStructure();
-        auto path_column = block_structure[0].type->createColumn();
-        auto is_passed_column = block_structure[1].type->createColumn();
-        auto message_column = block_structure[2].type->createColumn();
+        auto processors = std::make_shared<Processors>();
+
+        std::vector<OutputPort *> worker_ports;
+
+        /// For verbose mode (`check_query_single_value_result = 0`)
+        /// if `max_threads` is not explicitly set, we will use single thread to have deterministic output order
```

Comment:
Maybe we need the opposite? Max threads by default and deterministic order for tests

Reply:
Let's have one CI run with results shuffling https://github.com/ClickHouse/ClickHouse/pull/53404/commits/7b55c68130bd87a6c153dc2a9b7b55c91bf2fed3 to be sure that I didn't forget to add `max_threads = 1` in any test. Once it's passed, I'll remove the commit from this PR

Reply:
CI passed @ https://github.com/ClickHouse/ClickHouse/commit/1e068640c8f9912017612b9f40bacbb9029c4546, except an error in debug build. I don't think we need to rerun with randomization, release tests result is enough, I'm going to revert result randomization and fix debug build